wallet/cardano-trezor-crypto-provider.js

```diff
@@ -0,0 +1,370 @@
+      unpackAddress(address, hdPassphrase)
+      return true
+    } catch (e) {
+      if (e.name === 'AddressDecodingException') {
+        return false
+      }
+
+      throw e
+    }
+  }
+
+  function deriveHdNode(childIndex) {
+    throw new Error('This operation is not supported on TrezorCryptoProvider!')
+  }
+
+  async function sign(message, derivationPath) {
+    return await new Promise((resolve, reject) => {
+      const messageToSign = Buffer.from(message, 'hex').toString('utf8')
+
+      // m/44'/1815'/0'/0/childIndex
+      const path = normalizePathsForTrezor([derivationPath])[0]
+
+      callTrezor((shouldRejectOnError) => {
+        TrezorConnect.adaSignMessage(path, messageToSign, (response) => {
+          if (response.success) {
+            resolve(new Buffer(response.signature, 'hex'))
+          } else {
+            if (shouldRejectOnError(response.error)) {
+              reject(response.error)
+            }
+          }
+        })
+      })
+    })
+  }
+
+  async function getRootHdPassphrase() {
+    if (!state.rootHdPassphrase) {
+      const rootHdPassphrase = (await deriveTrezorXpub([])).root_hd_passphrase
+      state.rootHdPassphrase = new Buffer(rootHdPassphrase, 'hex')
+    }
+
+    return state.rootHdPassphrase
+  }
+
+  async function getDerivationPathFromAddress(address) {
+    const cachedAddress = Object.values(state.derivedAddresses).find((record) => {
+      if (record.address === address) {
+        return true
+      }
+
+      return false
+    })
+
+    if (cachedAddress) {
+      return cachedAddress.derivationPath
+    } else {
+      return unpackAddress(address, await getRootHdPassphrase()).derivationPath
+    }
+  }
+
+  async function _prepareInput(input) {
+    const data = {
+      tx_hash: input.txHash,
+      output_index: input.outputIndex,
+      type: 0,
+    }
+
+    const derivationPath = await getDerivationPathFromAddress(input.utxo.address)
+    data.address_n = normalizePathsForTrezor([derivationPath])[0]
+
+    return data
+  }
+
+  async function _prepareOutput(output) {
+    const data = {
+      amount: output.coins,
+    }
+
+    if (output.isChange) {
+      const derivationPath = await getDerivationPathFromAddress(output.address)
+      data.address_n = normalizePathsForTrezor([derivationPath])[0]
+    } else {
+      data.address = output.address
+    }
+
+    return data
+  }
+
+  async function signTx(unsignedTx) {
+    const inputs = []
+    for (const input of unsignedTx.inputs) {
+      inputs.push(await _prepareInput(input))
+    }
+
+    const transactions = []
+    for (const input of inputs) {
+      const transaction = await blockchainExplorer.fetchTxRaw(input.tx_hash)
+      transactions.push(transaction)
+    }
+
+    const outputs = []
+    for (const output of unsignedTx.outputs) {
+      const data = await _prepareOutput(output)
+      outputs.push(data)
+    }
+
+    const attributes = []
+
+    return await new Promise((resolve, reject) => {
```

Comment:
this await is probably redundant. The calling function should already call await on this promise